### PR TITLE
SWITCH: add nintendo switch support (updated)

### PR DESCRIPTION
**Pull request by rsn8887**
- This PR is based on @Cpasjuste's branch and PR #1341. I just rebased on latest master, resolved the  merge conflict in configure, added the needed _sdl=auto to configure, then squashed it for clarity.
- My additional commit https://github.com/scummvm/scummvm/pull/1619/commits/254f223559037053d2ac5cf2464fce29d3d9da18 implements the suggestions from the review by @ccawley2011 (see comments and review in #1341).
- For the reasons outlined in https://github.com/scummvm/scummvm/pull/1341#discussion_r276461401 I did not merge the event managers of the Vita (PSP2) platform and Switch platform together.
- As per @digitall's suggestion in https://github.com/scummvm/scummvm/pull/1341#issuecomment-433136766, I also added the README.SWITCH file in https://github.com/scummvm/scummvm/pull/1619/commits/254f223559037053d2ac5cf2464fce29d3d9da18.
- I am opening this new PR because @cpasjuste told me on Discord that he won't have time to work on this anymore.

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -3408,6 +3435,19 @@ if test -n "$_host"; then
 			_mt32emu=no
 			_vkeybd=yes
 			;;
+		switch)
+			_backend="switch"
+			_opengl_mode=gles2
+			_build_scalers=yes
```
Comment (ccawley2011) [inline configure]: `_build_scalers`, `_build_hq_scalers`, `_dynamic_modules`, `_mt32emu`, `_vorbis`, `_tremor` and `_eventrec` don't need to be set here as they are already set appropriately by default.

File: backends/platform/sdl/switch/switch-main.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include <switch.h>
+
+#define FORBIDDEN_SYMBOL_EXCEPTION_stdout
+#define FORBIDDEN_SYMBOL_EXCEPTION_stderr
```
Comment (ccawley2011) [inline backends/platform/sdl/switch/switch-main.cpp]: These two defines aren't needed anymore.

File: backends/platform/sdl/switch/switch.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#define FORBIDDEN_SYMBOL_EXCEPTION_printf
+#define FORBIDDEN_SYMBOL_EXCEPTION_unistd_h
```
Comment (ccawley2011) [inline backends/platform/sdl/switch/switch.cpp]: This define isn't needed either.

File: backends/platform/sdl/switch/switch.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+#include <sys/stat.h>
```
Comment (ccawley2011) [inline backends/platform/sdl/switch/switch.cpp]: This include isn't needed anymore.

**Comment by ccawley2011** [comment]
> For the reasons outlined in #1341 (comment) I did not merge the event managers of the Vita (PSP2) platform and Switch platform together.

I still think that something should be done to this end. The main issue here is duplicating large amounts of code like this makes maintenance somewhat problematic. This isn't a massive issue at the moment, since currently only the Vita and Switch backend uses it, and you're the maintainer for both of them, however this can potentially cause a bigger issue if the code is duplicated again for more potential platforms, it'll become harder to make further changes regarding touch input for the SDL backend.

A similar problem is present with the GPH, Dingux and LinuxMoto graphics managers, which all use identical-but-not-quite copies of the downscaling code, which makes it harder to adapt them for changes like PR #271.

A potential solution would be to create a touchscreen event manager subclassing the SDL event manager containing most of the touchscreen code, with the addition of some easily overridable functions to determine what mode each touch panel should use (plus anything I might have missed), and have the Vita and Switch event managers subclass that instead of the regular SDL event manager.

What do you think?

File: backends/platform/sdl/switch/switch.mk [hunk 5]
```diff
@@ -0,0 +1,19 @@
+scummvm.nro: $(EXECUTABLE)
+	mkdir -p ./switch_release/scummvm/data
+	mkdir -p ./switch_release/scummvm/doc
+	nacptool --create "ScummVM" "Cpasjuste" "2.12" ./switch_release/scummvm.nacp
```
[the remark this answers is not in the document]
Comment (rsn8887) [reply backends/platform/sdl/switch/switch.mk]: Fixed in 4fce992.

File: backends/platform/sdl/switch/switch-main.cpp (hunk 2 above)
Comment (rsn8887) [reply backends/platform/sdl/switch/switch-main.cpp]: fixed in 4fce992

File: configure (hunk 1 above)
Comment (rsn8887) [reply configure]: fixed in 4fce992

File: configure [hunk 6]
```diff
@@ -1680,6 +1681,16 @@ samsungtv)
...
 	;;
+switch)
+	_host_os=switch
+	_host_cpu=aarch64
+	_host_alias=aarch64-none-elf
+	test "x$prefix" = xNONE && prefix=.
+	datarootdir='${prefix}/data'
+	datadir='${datarootdir}'
+	docdir='${prefix}/doc'
+	PKG_CONFIG_LIBDIR=$DEVKITPRO/portlibs/switch/lib/pkgconfig
```
[the remark this answers is not in the document]
Comment (rsn8887) [reply configure]: fixed in 4fce992

File: backends/platform/sdl/switch/switch.cpp (hunk 3 above)
Comment (rsn8887) [reply backends/platform/sdl/switch/switch.cpp]: fixed in 4fce992

File: backends/platform/sdl/switch/switch.cpp (hunk 4 above)
Comment (rsn8887) [reply backends/platform/sdl/switch/switch.cpp]: fixed in 4fce992

**Comment by rsn8887** [comment]
I think that is a great idea, however, I just don't have the time and willpower to refactor and then re-debug this code anytime soon. Basically, this touch code took me forever to get just right and is now used in many other ports by me and other homebrew devs on Vita and Switch. A refactor would require debugging this thing again.

Maybe this could be deferred to a later time, after the Switch is merged, but before a third platform would start to make use of this code?

**Comment by criezy** [comment]
Adding a touchscreen event manager subclassing the SDL one to factorize some code sounds like a good idea, but it may indeed be outside the scope of this PR.

**Comment by Cpasjuste** [comment]
Thanks for your time @rsn8887 :)

**Comment by rsn8887** [comment]
You're welcome. Thank YOU for all your work on this.

**Comment by Mataniko** [comment]
Thank you both!

**Comment by bluegr** [comment]
Overall, the changes are good :)

Regarding the touch screen refactoring: work has been done on this as part of GSoC, and the results can be found in PR #401. You're welcome to have a look at the PR, and help to merge it in our main code base!

**Comment by bluegr** [comment]
Overall, great work, and thanks for addressing the issues of the original PR. The refactoring needed to reduce the code duplication of the events manager code can be done in-tree.

The Switch code is in a very good shape at the moment, it has been open for 5 days, and it has received good reviews already.  The original PR has been open for a long time, so most of the code has already been reviewed.

Thus, this can be merged, since there haven’t been any objections or further comments.

Thanks again for your work! Merging

**Comment by Cpasjuste** [comment]
Finally, thanks to all of you. It's great to see this merged. I'm very bad at finishing things :/

**Comment by sev-** [comment]
Thank you for your work, @Cpasjuste. I hope that you will have time for providing the release binaries :). We will start the release cycle soon.

**Comment by rsn8887** [comment]
I can do that too if needed no problem

**Pull request merged**